### DISCUSSION TITLE
Using Clang.jl for wrapping Objective C code

### DISCUSSION THREAD

**Discussion by maleadt** [JuliaInterop/Clang.jl#416] [General]
I would want to use Clang.jl for parsing Objective C headers and generating wrappers that use the `@objc` macro from ObjectiveC.jl (which is still under development). So going from:

```objc
@interface NSConditionLock : NSObject <NSLocking> {
@private
    void *_priv;
}

- (instancetype)initWithCondition:(NSInteger)condition NS_DESIGNATED_INITIALIZER;

@property (readonly) NSInteger condition;
- (void)lockWhenCondition:(NSInteger)condition;
- (BOOL)tryLock;
- (BOOL)tryLockWhenCondition:(NSInteger)condition;
- (void)unlockWithCondition:(NSInteger)condition;
- (BOOL)lockBeforeDate:(NSDate *)limit;
- (BOOL)lockWhenCondition:(NSInteger)condition beforeDate:(NSDate *)limit;

@property (nullable, copy) NSString *name API_AVAILABLE(macos(10.5), ios(2.0), watchos(2.0), tvos(9.0));

@end
```

... to something like:

```julia
abstract type NSConditionLock <: NSObject
end

# example of a class method
function initWithCondition(condition)
    @objc NSConditionLock initWithCondition(condition::NSInteger)::NSConditionLock
end

# example of an instance method
function tryLock(self::NSConditionLock)
    @objc self::NSConditionLock tryLock()::Bool
end
```

... or something like it. I'm still figuring out how I want the `@objc` macro to look like. Currently it looks completely different, e.g. for the above two functions it'd do `@objc [NSConditionLock initWithCondition:condition]` and `@objc [self tryLock]`, relying on run-time introspection to figure out the types of the arguments and return value, but we should be able to parse all that from the headers.

Anyway, what would be the recommended way to tackle this? Writing a custom generator, I presume? Or can the existing C generator be repurposed? If not, is there an example of a minimal generator to start with (the C one is pretty huge)?

**Comment by Gnimuc** [comment]
I never used Objective C, but if you only need to wrap interfaces(functions) with built-in types. Extracting required info from ObjC headers should not be very hard.

We have a step-by-step tutorial about how to use Clang APIs:
https://juliainterop.github.io/Clang.jl/dev/tutorial/

and a very simple example of how to extract info from C++ headers:

https://github.com/JuliaInterop/Clang.jl/blob/43fa7c5fc50c6b75034cb478c4a8518062c76d8a/examples/generator.jl

ObjC should be similar. 

The source of C generator is huge because it's tricky to handle forward declaration and nested/anonymous structs/unions. If you need to wrap those C code written in ObjC, you could define a new context and reuse those passes on demand:
https://github.com/JuliaInterop/Clang.jl/blob/5cc22528b48da6883894bda1382526565db882e3/src/generator/context.jl#L80
